unison-cli/src/Unison/Codebase/Editor/HandleInput/NamespaceDependencies.hs

```diff
@@ -0,0 +1,104 @@
+-- returns a map of dependencies which do not have a name within the current namespace,
+-- alongside the names of all of that thing's dependents.
+--
+-- This is non-transitive, i.e. only the first layer of external dependencies is returned.
+--
+-- So if my namespace depends on .base.Bag.map; which depends on base.Map.mapKeys, only
+-- .base.Bag.map is returned unless some other definition inside my namespace depends
+-- on .base.Bag.map directly.
+namespaceDependencies :: forall m i v. Branch0 m -> Action m i v (Map Referent (Set Name))
+namespaceDependencies branch = do
+  let refToDeps :: Reference.Id -> Action m i v (Map Referent (Set Name))
+      refToDeps refId = do
+        dependencies <- fmap Referent.fromReference . Set.toList <$> eval (GetDependencies refId)
+        ( dependencies
+            & fmap \dep -> Map.singleton dep (localNameByReference (Reference.fromId refId))
+          )
+          & Map.unionsWith (<>)
+          & pure
+  let allDependenciesOf :: Set Reference -> Action m i v (Map Referent (Set Name))
+      allDependenciesOf refs =
+        refs
+          & Set.toList
+          & mapMaybe Reference.toId
+          & traverse refToDeps
+          <&> Map.unionsWith (<>)
+
+  typeDeps <- allDependenciesOf currentBranchTypes
+  termDeps <- allDependenciesOf (Set.map Referent.toReference currentBranchTermsAndConstructors)
+  let dependenciesToDependents :: Map Referent (Set Name)
+      dependenciesToDependents =
+        Map.unionsWith (<>) [typeDeps, termDeps, externalMetadata]
+  let onlyExternalDeps :: Map Referent (Set Name)
+      onlyExternalDeps =
+        Map.filterWithKey
+          (\k _ -> k `Set.notMember` typeAndTermRefsInCurrentBranch)
+          dependenciesToDependents
+  externalConstructors :: Map Referent (Set Name) <-
+    Map.unions . concat
+      <$> ( for (Map.toList onlyExternalDeps) $ \case
+              (ref, deps) -> do
+                constructors <- eval (ConstructorsOfType (Referent.toReference ref))
+                let externalConstrs = constructors `Set.difference` currentBranchTermsAndConstructors
+                pure $
+                  [ Map.singleton ref deps,
+                    Map.fromListWith (<>) ((,deps) <$> Set.toList externalConstrs)
+                  ]
+          )
+  let allDependenciesToDependents = Map.unionWith (<>) externalConstructors onlyExternalDeps
+  pure allDependenciesToDependents
+  where
+    currentBranchTermsAndConstructors :: Set Referent
+    currentBranchTermsAndConstructors = Relation.dom (Branch.deepTerms branch)
+    localNameByReference :: Reference -> (Set Name)
+    localNameByReference ref = Relation.lookupDom (Referent.fromReference ref) (Branch.deepTerms branch)
+    currentBranchTypes :: Set Reference
+    currentBranchTypes = Relation.dom (Branch.deepTypes branch)
+    typeAndTermRefsInCurrentBranch :: Set Referent
+    typeAndTermRefsInCurrentBranch =
+      Set.map Referent.fromReference (Relation.dom (Branch.deepTypes branch))
```

Comment:
I see; it does look here to be putting type references into a `Referent.Ref`, which is danger zone.  I suggest using `LabeledDependency` here instead, if you need both term referents and type references in the same collection.